src/graph.js

```diff
@@ -26,18 +26,21 @@ const chainP = R.curry((fn, iterableP) => {
   )(iterableP);
 });
 
-function walk(file) {
+const walk = R.curry((visited, file) => {
   const basedir = dirname(file);
-  return R.pipeP(resolve,
-    esDeps,
-    R.map(R.pipe(
-      join(basedir),
-      resolveFile
-    )),
-    chainP(walk),
-    R.prepend(file)
+  return R.unless(
+    R.contains(R.__, visited),
+    R.pipeP(resolve,
+      esDeps,
+      R.map(R.pipe(
+        join(basedir),
+        resolveFile
+      )),
+      chainP(walk(R.append(file, visited))),
```

Comment:
i had similar probles in [es-deps](https://github.com/iamstarkov/es-deps-from-string/blob/master/index.js#L27-L33)